circle/src/domain.rs

```diff
@@ -0,0 +1,355 @@
+use alloc::vec;
+use alloc::vec::Vec;
+
+use itertools::Itertools;
+use p3_commit::{LagrangeSelectors, PolynomialSpace};
+use p3_field::extension::{Complex, ComplexExtendable};
+use p3_field::{batch_multiplicative_inverse, AbstractField, ExtensionField};
+use p3_matrix::dense::RowMajorMatrix;
+use p3_matrix::Matrix;
+use p3_util::{log2_ceil_usize, log2_strict_usize};
+use tracing::instrument;
+
+use crate::util::{point_to_univariate, s_p_at_p, univariate_to_point, v_0, v_n};
+
+/// X is generator, O is the first coset, goes counterclockwise
+/// ```text
+///   O X .
+///  .     .
+/// .       O <- start = shift
+/// .   .   - (1,0)
+/// O       .
+///  .     .
+///   . . O
+/// ```
+///
+/// For ordering reasons, the other half will start at gen / shift:
+/// ```text
+///   . X O  <- start = gen/shift
+///  .     .
+/// O       .
+/// .   .   - (1,0)
+/// .       O
+///  .     .
+///   O . .
+/// ```
+///
+/// The full domain is the interleaving of these two cosets
+#[derive(Copy, Clone, PartialEq, Eq, Debug)]
+pub struct CircleDomain<F> {
```

Comment:
I would just drop the term twin-coset at some place, for easier reference to the paper.